File: SunnyWeather-pro/src/main/java/com/example/sunnyweather_pro/logic/network/SunnyWeatherNetwork.java
package com.example.sunnyweather_pro.logic.network;

import android.util.Log;
import android.widget.Toast;

import androidx.annotation.NonNull;

import com.example.sunnyweather_pro.MyApplication;
import com.example.sunnyweather_pro.logic.model.DailyResponse;
import com.example.sunnyweather_pro.logic.model.PlaceResponse;
import com.example.sunnyweather_pro.logic.model.RealtimeResponse;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

/*定义一个统一的网络数据源访问入口，对所有网络请求的API进行封装*/
public class SunnyWeatherNetwork {

    private static PlaceResponse placeResponse;
    private static RealtimeResponse realtimeResponse;
    private static DailyResponse dailyResponse;

    //创建Service接口的动态代理对象
    private static final PlaceService placeService = ServiceCreator.create(PlaceService.class);
    private static final WeatherService weatherService = ServiceCreator.create(WeatherService.class);

    //搜索城市地址
    public static PlaceResponse searchPlaces(String query){
        //该方法是一个异步请求，用于搜索与查询字符串 query 相关的城市地址
        Call<PlaceResponse> call = placeService.searchPlaces(query);
        call.enqueue(new Callback<PlaceResponse>() {
            @Override
            public void onResponse(@NonNull Call<PlaceResponse> call, @NonNull Response<PlaceResponse> response) {
                Log.d("boge",response.toString());
                if (response.body() != null){
                    placeResponse = response.body();
                } else Toast.makeText(MyApplication.context, "响应体为空", Toast.LENGTH_SHORT).show();
            }
            @Override
            public void onFailure(@NonNull Call<PlaceResponse> call, @NonNull Throwable t) {
                t.printStackTrace();
                Toast.makeText(MyApplication.context, t.getMessage(), Toast.LENGTH_SHORT).show();
            }
        });
        return placeResponse;
    }

    //获取实时天气数据
    public static RealtimeResponse getRealtimeWeather(String lng,String lat){
        Call<RealtimeResponse> call = weatherService.getRealtimeWeather(MyApplication.TOKEN, lng, lat);
        call.enqueue(new Callback<RealtimeResponse>() {
            @Override
            public void onResponse(@NonNull Call<RealtimeResponse> call, @NonNull Response<RealtimeResponse> response) {
                if (response.body() != null){
                    realtimeResponse = response.body();
                    //Log.d("boge","realtime = " + realtimeResponse.result.realtime.toString());
                } else Toast.makeText(MyApplication.context, "响应体为空", Toast.LENGTH_SHORT).show();
            }

            @Override
            public void onFailure(@NonNull Call<RealtimeResponse> call, @NonNull Throwable t) {
                t.printStackTrace();
                Toast.makeText(MyApplication.context, t.getMessage(), Toast.LENGTH_SHORT).show();
            }
        });
        return realtimeResponse;
    }

    //用于获取每日天气数据
    public static DailyResponse getDailyWeather(String lng, String lat){
        Call<DailyResponse> call = weatherService.getDailyWeather(MyApplication.TOKEN, lng, lat);
        call.enqueue(new Callback<DailyResponse>() {
            @Override
            public void onResponse(@NonNull Call<DailyResponse> call, @NonNull Response<DailyResponse> response) {
                if (response.body() != null){
                    dailyResponse = response.body();
                    //Log.d("boge","daily = " + dailyResponse.result.daily.toString());
                } else Toast.makeText(MyApplication.context, "响应体为空", Toast.LENGTH_SHORT).show();
            }

            @Override
            public void onFailure(@NonNull Call<DailyResponse> call, @NonNull Throwable t) {
                t.printStackTrace();
                Toast.makeText(MyApplication.context, t.getMessage(), Toast.LENGTH_SHORT).show();
            }
        });
        return dailyResponse;
    }

}

File: SunnyWeather-kotlin/src/main/java/com/driverskr/sunnyweather_kotlin/logic/network/SunnyWeatherNetwork.kt
package com.driverskr.sunnyweather_kotlin.logic.network

import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import kotlin.coroutines.resume
import kotlin.coroutines.resumeWithException
import kotlin.coroutines.suspendCoroutine

/**
 * @Author: driverSkr
 * @Time: 2023/11/6 15:18
 * @Description: 统一的网络数据源访问入口$
 */
object SunnyWeatherNetwork {

    private val placeService = ServiceCreator.create<PlaceService>()

    suspend fun searchPlaces(query: String) = placeService.searchPlaces(query).await()

    /**
     * 当外部调用SunnyWeatherNetwork的searchPlaces()函数时，Retrofit就会立即发起网络请求，同时当前的协程也会被阻塞住。
     * 直到服务器响应我们的请求之后，await()函数会将解析出来的数据模型对象取出并返回，同时恢复当前协程的执行，
     * searchPlaces()函数在得到await()函数的返回值后会将该数据再返回到上一层。
     */
    private suspend fun <T> Call<T>.await(): T {
        return suspendCoroutine { continuation ->
            enqueue(object : Callback<T> {
                override fun onResponse(call: Call<T>, response: Response<T>) {
                    val body = response.body()
                    if (body != null) continuation.resume(body)
                    else continuation.resumeWithException(RuntimeException("response body is null"))
                }
                override fun onFailure(call: Call<T>, t: Throwable) {
                    continuation.resumeWithException(t)
                }
            })
        }
    }
}